internal/cimfs/cim_writer_windows.go

```diff
@@ -0,0 +1,263 @@
+	err := c.closeStream()
+	if err != nil {
+		return err
+	}
+	fileMetadata := &winapi.CimFsFileMetadata{
+		Attributes:     info.FileAttributes,
+		FileSize:       fileSize,
+		CreationTime:   info.CreationTime,
+		LastWriteTime:  info.LastWriteTime,
+		ChangeTime:     info.ChangeTime,
+		LastAccessTime: info.LastAccessTime,
+	}
+	if len(securityDescriptor) == 0 {
+		// Passing an empty security descriptor creates a CIM in a weird state.
+		// Pass the NULL DACL.
+		securityDescriptor = nullSd
+	}
+	fileMetadata.SecurityDescriptorBuffer = unsafe.Pointer(&securityDescriptor[0])
+	fileMetadata.SecurityDescriptorSize = uint32(len(securityDescriptor))
+	if len(reparseData) > 0 {
+		fileMetadata.ReparseDataBuffer = unsafe.Pointer(&reparseData[0])
+		fileMetadata.ReparseDataSize = uint32(len(reparseData))
+	}
+	if len(extendedAttributes) > 0 {
+		fileMetadata.ExtendedAttributes = unsafe.Pointer(&extendedAttributes[0])
+		fileMetadata.EACount = uint32(len(extendedAttributes))
+	}
+	err = winapi.CimCreateFile(c.handle, path, fileMetadata, &c.activeStream)
+	if err != nil {
+		return &PathError{Cim: c.name, Op: "addFile", Path: path, Err: err}
+	}
+	c.activeName = path
+	if info.FileAttributes&(windows.FILE_ATTRIBUTE_DIRECTORY) == 0 {
+		c.activeLeft = fileSize
+	}
+	return nil
+}
+
+// This is a helper function which reads the file on host at path `hostPath` and adds it
+// inside the cim at path `pathInCim`. If a file already exists inside cim at path
+// `pathInCim` it will be overwritten.
+func (c *CimFsWriter) AddFileFromPath(pathInCim, hostPath string, securityDescriptor []byte, extendedAttributes []byte, reparseData []byte) error {
+	f, err := os.Open(hostPath)
+	if err != nil {
+		return errors.Wrapf(err, "AddFileFromPath, can't open file: %s", hostPath)
+	}
+	defer f.Close()
+
+	basicInfo, err := winio.GetFileBasicInfo(f)
+	if err != nil {
+		return errors.Wrapf(err, "AddFileFromPath, failed to get file info for %s", hostPath)
+	}
+
+	replaceData, err := ioutil.ReadFile(hostPath)
+	if err != nil {
+		return errors.Wrapf(err, "AddFileFromPath, unable to read file %s", hostPath)
+	}
+	if err := c.AddFile(pathInCim, basicInfo, int64(len(replaceData)), securityDescriptor, extendedAttributes, reparseData); err != nil {
+		return err
+	}
+
+	if _, err := c.Write(replaceData); err != nil {
+		return &PathError{Cim: c.name, Op: "write", Path: c.activeName, Err: err}
+	}
+	return nil
+}
+
+// Write writes bytes to the active stream.
+func (c *CimFsWriter) Write(p []byte) (int, error) {
+	if c.activeStream == 0 {
+		return 0, errors.New("no active stream")
+	}
+	if int64(len(p)) > c.activeLeft {
+		return 0, &PathError{Cim: c.name, Op: "write", Path: c.activeName, Err: errors.New("wrote too much")}
+	}
+	err := winapi.CimWriteStream(c.activeStream, uintptr(unsafe.Pointer(&p[0])), uint32(len(p)))
+	if err != nil {
+		err = &PathError{Cim: c.name, Op: "write", Path: c.activeName, Err: err}
+		return 0, err
+	}
+	c.activeLeft -= int64(len(p))
```

Comment:
Hmm, good point. Current containerd client writes files one by one so this is not a problem for now, but it could be a problem if we add new code that uses this writer. I can add locks around all CimFS API calls, what do you think?